tests/compas/robots/test_model.py

```diff
@@ -116,21 +117,61 @@ def test_robot_urdf_namespaces_to_string():
     assert isinstance(r, RobotModel)
     assert r.name == 'Acrobot'
 
 
+def test_robot_link_nameless_is_allowed_with_custom_namespace():
+    r = RobotModel.from_urdf_string(
+        """<?xml version="1.0" encoding="UTF-8"?><robot xmlns:namelesslinks="https://somewhere.over.the.rainbow" name="NamelessLinkRobot"><namelesslinks:link/></robot>""")
+    assert isinstance(r, RobotModel)
+    assert r.name == 'NamelessLinkRobot'
+
+def test_link_nameless_raises_if_no_custom_namespace():
+    with pytest.raises(Exception):
+        r = RobotModel.from_urdf_string(
+            """<?xml version="1.0" encoding="UTF-8"?><robot name="NamelessLinkRobot"><link/></robot>""")
+
+def test_robot_default_namespace_creates_box_shape_based_on_tagname():
+    r = RobotModel.from_urdf_string(
+        """<?xml version="1.0"?><robot xmlns="https://drake.mit.edu" name="Acrobot"><link name="base_link"><visual><geometry><box size="0.2 0.2 0.2"/></geometry></visual></link></robot>""")
+    assert r.name == 'Acrobot'
+    assert r.links[0].name == 'base_link'
+    assert isinstance(r.links[0].visual[0].geometry.shape, Box)
+
 def test_robot_default_namespace_to_string():
     r = RobotModel.from_urdf_string(
         """<?xml version="1.0" encoding="UTF-8"?><robot xmlns="https://drake.mit.edu" name="Acrobot"><frame/></robot>""")
     urdf_string = URDF.from_robot(r).to_string(prettify=True)
     assert b'xmlns="https://drake.mit.edu"' in urdf_string
     assert b'<frame' in urdf_string or b'<ns0:frame' in urdf_string
 
+def test_robot_with_default_nested_namespaces():
+    r = RobotModel.from_urdf_string(
+        """<?xml version="1.0" encoding="UTF-8"?><robot xmlns="https://ethz.ch" name="Acrobot"><link xmlns="https://ita.ethz.ch" name="test"/></robot>""")
+
+    urdf = URDF.from_robot(r)
+    assert urdf.robot.attr['xmlns'] == 'https://ethz.ch'
+    assert urdf.robot.links[0].attr['xmlns'] == 'https://ita.ethz.ch'
+
+def test_robot_with_default_nested_namespaces_to_string():
+    r = RobotModel.from_urdf_string(
+        """<?xml version="1.0" encoding="UTF-8"?><robot xmlns="https://ethz.ch" name="Acrobot"><link xmlns="https://ita.ethz.ch" name="test"><visual><geometry><box size="0.2 0.2 0.2"/></geometry></visual></link></robot>""")
+    urdf_string = URDF.from_robot(r).to_string(prettify=False)
+    assert re.search(b'<robot(.*?)(xmlns="https://ethz.ch")(.*?)>', urdf_string)
+    assert re.search(b'<link(.*?)(xmlns="https://ita.ethz.ch")(.*?)>', urdf_string)
+
+
+def test_robot_with_prefixed_nested_namespaces_to_string():
+    r = RobotModel.from_urdf_string(
+        """<?xml version="1.0" encoding="UTF-8"?><robot xmlns="https://ethz.ch" name="Acrobot"><link xmlns:custom="https://ita.ethz.ch" name="test"><custom:visual/></link></robot>""")
+    urdf_string = URDF.from_robot(r).to_string(prettify=False)
+    assert b'xmlns="https://ethz.ch"' in urdf_string
+    assert b'xmlns:ns0="https://ita.ethz.ch"' in urdf_string
+    assert b'<ns0:visual' in urdf_string
```

Comment:
i'm surprised these pass in ipy.  maybe the second `assert` of `test_robot_default_namespace_to_string` is too verbose.

Reply:
indeed! fixed at https://github.com/compas-dev/compas/pull/686/commits/a801ac1c9a28f5df77ca59b9d0e11a43fb7d8101